File: sdk/flutter/revora_flutter/ios/Classes/RevoraFlutterPlugin.swift
import Flutter
import UIKit

public class RevoraFlutterPlugin: NSObject, FlutterPlugin {
    var owner: String = ""
    var repo: String = ""
    var publicKey: SecKey?

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "revora_flutter", binaryMessenger: registrar.messenger())
        let instance = RevoraFlutterPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "initialize":
            if let args = call.arguments as? [String: Any] {
                owner = args["owner"] as? String ?? ""
                repo = args["repo"] as? String ?? ""
                if let keyStr = args["publicKey"] as? String {
                    publicKey = loadPublicKey(keyStr)
                }
            }
            result(nil)
        case "checkForUpdate":
            checkForUpdate { available in result(available) }
        case "applyUpdate":
            applyUpdate { applied in result(applied) }
        case "getCurrentVersion":
            result(getCurrentVersion())
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func loadPublicKey(_ keyStr: String) -> SecKey? {
        guard let data = Data(base64Encoded: keyStr) else { return nil }
        let attributes: [String: Any] = [
            kSecAttrKeyType as String: kSecAttrKeyTypeRSA,
            kSecAttrKeyClass as String: kSecAttrKeyClassPublic
        ]
        return SecKeyCreateWithData(data as CFData, attributes as CFDictionary, nil)
    }

    private func getCurrentVersion() -> String {
        return Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String ?? "0.0.0"
    }

    private func checkForUpdate(completion: @escaping (Bool) -> Void) {
        let currentVersion = getCurrentVersion()
        let url = URL(string: "https://api.github.com/repos/\(owner)/\(repo)/releases/latest")!
        URLSession.shared.dataTask(with: url) { data, _, error in
            guard let data = data, error == nil,
                  let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let tagName = json["tag_name"] as? String,
                  tagName != currentVersion else {
                completion(false)
                return
            }
            if let assets = json["assets"] as? [[String: Any]] {
                for asset in assets {
                    if let name = asset["name"] as? String,
                       let downloadUrl = asset["browser_download_url"] as? String {
                        switch name {
                        case "update.patch":
                            UserDefaults.standard.set(downloadUrl, forKey: "revora_patch_url")
                        case "manifest.json":
                            UserDefaults.standard.set(downloadUrl, forKey: "revora_manifest_url")
                        case "manifest.json.sig":
                            UserDefaults.standard.set(downloadUrl, forKey: "revora_signature_url")
                        default: break
                        }
                    }
                }
                completion(true)
            } else {
                completion(false)
            }
        }.resume()
    }

    private func applyUpdate(completion: @escaping (Bool) -> Void) {
        guard let manifestUrl = UserDefaults.standard.string(forKey: "revora_manifest_url"),
              let patchUrl = UserDefaults.standard.string(forKey: "revora_patch_url"),
              let signatureUrl = UserDefaults.standard.string(forKey: "revora_signature_url"),
              let publicKey = publicKey else {
            completion(false)
            return
        }

        // Download manifest and signature
        downloadText(url: URL(string: manifestUrl)!) { manifestData in
            guard let manifestData = manifestData else { completion(false); return }
            self.downloadText(url: URL(string: signatureUrl)!) { signatureB64 in
                guard let signatureB64 = signatureB64,
                      let signatureData = Data(base64Encoded: signatureB64) else { completion(false); return }

                var error: Unmanaged<CFError>?
                let valid = SecKeyVerifySignature(publicKey, .rsaSignatureMessagePKCS1v15SHA256, manifestData as CFData, signatureData as CFData, &error)
                if !valid { completion(false); return }

                // Download patch
                self.downloadFile(url: URL(string: patchUrl)!) { localURL in
                    guard let localURL = localURL else { completion(false); return }
                    let pendingDir = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask).first!
                        .appendingPathComponent("revora_pending")
                    try? FileManager.default.createDirectory(at: pendingDir, withIntermediateDirectories: true)
                    let dest = pendingDir.appendingPathComponent("update.patch")
                    try? FileManager.default.copyItem(at: localURL, to: dest)
                    UserDefaults.standard.set(true, forKey: "revora_pending_restart")
                    completion(true)
                }
            }
        }
    }

    private func downloadText(url: URL, completion: @escaping (Data?) -> Void) {
        URLSession.shared.dataTask(with: url) { data, _, _ in completion(data) }.resume()
    }

    private func downloadFile(url: URL, completion: @escaping (URL?) -> Void) {
        URLSession.shared.downloadTask(with: url) { localURL, _, _ in completion(localURL) }.resume()
    }
}

File: sdk/flutter/revora_flutter/android/src/main/kotlin/com/example/revora_flutter/RevoraFlutterPlugin.kt
package com.example.revora_flutter

import android.content.Context
import android.content.SharedPreferences
import android.util.Base64
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import kotlinx.coroutines.*
import org.json.JSONObject
import java.io.File
import java.io.FileOutputStream
import java.net.HttpURLConnection
import java.net.URL
import java.security.KeyFactory
import java.security.PublicKey
import java.security.Signature
import java.security.spec.X509EncodedKeySpec

class RevoraFlutterPlugin : FlutterPlugin, MethodChannel.MethodCallHandler {
    private lateinit var channel: MethodChannel
    private lateinit var context: Context
    private var owner = ""
    private var repo = ""
    private var publicKey: PublicKey? = null
    private val scope = CoroutineScope(Dispatchers.IO)

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        context = binding.applicationContext
        channel = MethodChannel(binding.binaryMessenger, "revora_flutter")
        channel.setMethodCallHandler(this)
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "initialize" -> {
                owner = call.argument<String>("owner") ?: ""
                repo = call.argument<String>("repo") ?: ""
                val keyStr = call.argument<String>("publicKey") ?: ""
                publicKey = loadPublicKey(keyStr)
                result.success(null)
            }
            "checkForUpdate" -> {
                scope.launch {
                    val updateAvailable = checkForUpdateAsync()
                    withContext(Dispatchers.Main) { result.success(updateAvailable) }
                }
            }
            "applyUpdate" -> {
                scope.launch {
                    val applied = applyUpdateAsync()
                    withContext(Dispatchers.Main) { result.success(applied) }
                }
            }
            "getCurrentVersion" -> {
                result.success(getCurrentVersion())
            }
            else -> result.notImplemented()
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }

    private fun loadPublicKey(keyStr: String): PublicKey? {
        return try {
            val keyBytes = Base64.decode(keyStr, Base64.DEFAULT)
            val spec = X509EncodedKeySpec(keyBytes)
            KeyFactory.getInstance("RSA").generatePublic(spec)
        } catch (e: Exception) { null }
    }

    private suspend fun checkForUpdateAsync(): Boolean {
        val currentVersion = getCurrentVersion()
        val url = URL("https://api.github.com/repos/$owner/$repo/releases/latest")
        return try {
            val conn = url.openConnection() as HttpURLConnection
            conn.connectTimeout = 5000
            conn.readTimeout = 5000
            val json = JSONObject(conn.inputStream.bufferedReader().readText())
            val latestTag = json.optString("tag_name", "0.0.0")
            if (latestTag != currentVersion) {
                val assets = json.getJSONArray("assets")
                for (i in 0 until assets.length()) {
                    val asset = assets.getJSONObject(i)
                    val name = asset.getString("name")
                    val downloadUrl = asset.getString("browser_download_url")
                    when (name) {
                        "update.patch" -> getPrefs().edit().putString("revora_patch_url", downloadUrl).apply()
                        "manifest.json" -> getPrefs().edit().putString("revora_manifest_url", downloadUrl).apply()
                        "manifest.json.sig" -> getPrefs().edit().putString("revora_signature_url", downloadUrl).apply()
                    }
                }
                true
            } else false
        } catch (e: Exception) { false }
    }

    private suspend fun applyUpdateAsync(): Boolean {
        val prefs = getPrefs()
        val manifestUrl = prefs.getString("revora_manifest_url", null) ?: return false
        val patchUrl = prefs.getString("revora_patch_url", null) ?: return false
        val signatureUrl = prefs.getString("revora_signature_url", null) ?: return false

        val manifestData = downloadString(manifestUrl) ?: return false
        val signatureB64 = downloadString(signatureUrl) ?: return false
        val signatureBytes = Base64.decode(signatureB64, Base64.DEFAULT)

        if (publicKey == null) return false
        val sig = Signature.getInstance("SHA256withRSA")
        sig.initVerify(publicKey)
        sig.update(manifestData.toByteArray())
        if (!sig.verify(signatureBytes)) return false

        val patchFile = File(context.cacheDir, "update.patch")
        downloadFile(patchUrl, patchFile) ?: return false

        val pendingDir = File(context.filesDir, "revora_pending")
        pendingDir.mkdirs()
        patchFile.copyTo(File(pendingDir, "update.patch"), overwrite = true)
        prefs.edit().putBoolean("revora_pending_restart", true).apply()
        return true
    }

    private fun downloadString(urlStr: String): String? {
        return try {
            val conn = URL(urlStr).openConnection() as HttpURLConnection
            conn.connectTimeout = 5000
            conn.inputStream.bufferedReader().readText()
        } catch (e: Exception) { null }
    }

    private fun downloadFile(urlStr: String, dest: File): File? {
        return try {
            val conn = URL(urlStr).openConnection() as HttpURLConnection
            conn.connectTimeout = 15000
            conn.readTimeout = 15000
            conn.inputStream.use { input ->
                FileOutputStream(dest).use { output -> input.copyTo(output) }
            }
            dest
        } catch (e: Exception) { null }
    }

    private fun getCurrentVersion(): String {
        return try {
            val pkgInfo = context.packageManager.getPackageInfo(context.packageName, 0)
            pkgInfo.versionName ?: "0.0.0"
        } catch (e: Exception) { "0.0.0" }
    }

    private fun getPrefs(): SharedPreferences {
        return context.getSharedPreferences("revora", Context.MODE_PRIVATE)
    }
}